File: scripts/ensure_planning.py
# File: `scripts/ensure_planning.py`

```python
#!/usr/bin/env python3
"""
Planning gate checker for ai-project-starter.

This script intentionally uses simple marker checks.
It does not use fuzzy AI validation.

It fails when required planning files are missing, empty, or still unresolved.
"""

from __future__ import annotations

from pathlib import Path
import re
import sys


ROOT = Path(__file__).resolve().parents[1]

REQUIRED_FILES = [
    "PROJECT-INTAKE.md",
    "AGENTS.md",
    "HARNESS-LOOP.md",
    "00-PLANNING/DECISIONS-TO-MAKE.md",
    "ARCHITECTURE.md",
    "TECH-SPEC.md",
    "TASKS.md",
    "AI-NOTES.md",
]

OPTIONAL_CONTEXT_FILES = [
    "DATABASE.md",
    "DATA-FLOW.md",
]

PLACEHOLDER_PATTERNS = [
    r"\bTODO\b",
    r"\bTBD\b",
    r"\bFILL\s+THIS\b",
    r"\bPLACEHOLDER\b",
    r"\bUNRESOLVED\b",
    r"_Describe",
    r"_One paragraph",
    r"___________",
]


def read_file(path: Path) -> str:
    """Return file content as text."""
    return path.read_text(encoding="utf-8", errors="replace")


def fail(message: str, failures: list[str]) -> None:
    """Add a failure message."""
    failures.append(f"- {message}")


def file_has_real_content(content: str, min_non_empty_lines: int = 5) -> bool:
    """Check whether a file has enough non-empty lines to be useful."""
    lines = [line.strip() for line in content.splitlines() if line.strip()]
    return len(lines) >= min_non_empty_lines


def contains_bad_placeholder(content: str) -> list[str]:
    """Return placeholder markers found in content."""
    found: list[str] = []

    for pattern in PLACEHOLDER_PATTERNS:
        if re.search(pattern, content, flags=re.IGNORECASE):
            found.append(pattern)

    return found


def check_required_files(failures: list[str]) -> None:
    """Check that required planning files exist and have content."""
    for relative_path in REQUIRED_FILES:
        path = ROOT / relative_path

        if not path.exists():
            fail(f"Missing required file: `{relative_path}`", failures)
            continue

        content = read_file(path)

        if not file_has_real_content(content):
            fail(f"`{relative_path}` appears empty or too thin.", failures)


def check_project_intake(failures: list[str]) -> None:
    """Check PROJECT-INTAKE.md has a completion marker."""
    path = ROOT / "PROJECT-INTAKE.md"

    if not path.exists():
        return

    content = read_file(path)

    if "COMPLETED: true" not in content:
        fail("`PROJECT-INTAKE.md` must include `COMPLETED: true` before coding.", failures)


def check_decision_gates(failures: list[str]) -> None:
    """Check DECISIONS-TO-MAKE.md has resolved gate markers."""
    path = ROOT / "00-PLANNING/DECISIONS-TO-MAKE.md"

    if not path.exists():
        return

    content = read_file(path)

    unresolved_markers = [
        "STATUS: UNRESOLVED",
        "STATUS: NEEDS_SHAUN",
        "STATUS: NEEDS_SENIOR_DEVELOPER",
        "NEEDS SHAUN",
        "NEEDS SENIOR DEVELOPER",
    ]

    for marker in unresolved_markers:
        if marker in content:
            fail(f"`DECISIONS-TO-MAKE.md` still contains blocker marker: `{marker}`", failures)

    resolved_count = content.count("STATUS: RESOLVED")

    if resolved_count < 5:
        fail(
            "`DECISIONS-TO-MAKE.md` should contain at least 5 `STATUS: RESOLVED` markers "
            "for the core planning gates.",
            failures,
        )


def check_tech_spec(failures: list[str]) -> None:
    """Check TECH-SPEC.md has basic stack markers."""
    path = ROOT / "TECH-SPEC.md"

    if not path.exists():
        return

    content = read_file(path)

    required_markers = [
        "Runtime:",
        "Language:",
        "Database:",
        "Dependencies:",
    ]

    for marker in required_markers:
        if marker not in content:
            fail(f"`TECH-SPEC.md` should include `{marker}`.", failures)


def check_architecture(failures: list[str]) -> None:
    """Check ARCHITECTURE.md has structure markers."""
    path = ROOT / "ARCHITECTURE.md"

    if not path.exists():
        return

    content = read_file(path)

    required_markers = [
        "System Overview",
        "Architecture Style",
        "Directory Structure",
    ]

    for marker in required_markers:
        if marker not in content:
            fail(f"`ARCHITECTURE.md` should include `{marker}`.", failures)


def check_task_slice(failures: list[str]) -> None:
    """Check TASKS.md names the first useful build slice."""
    path = ROOT / "TASKS.md"

    if not path.exists():
        return

    content = read_file(path)

    if "FIRST BUILD SLICE:" not in content:
        fail("`TASKS.md` must include `FIRST BUILD SLICE:` before coding.", failures)


def main() -> int:
    """Run planning checks."""
    failures: list[str] = []

    check_required_files(failures)
    check_project_intake(failures)
    check_decision_gates(failures)
    check_tech_spec(failures)
    check_architecture(failures)
    check_task_slice(failures)

    if failures:
        print("Planning check failed.\n")
        print("Fix these items before coding or merging:\n")
        print("\n".join(failures))
        return 1

    print("Planning check passed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
```

---

# File: `.github/workflows/planning-check.yml`

```yaml
name: planning/check

on:
  pull_request:
    branches:
      - main
  workflow_dispatch:

jobs:
  planning-check:
    name: planning/check
    runs-on: ubuntu-latest

    steps:
      - name: Check out repository
        uses: actions/checkout@v4

      - name: Set up Python
        uses: actions/setup-python@v5
        with:
          python-version: "3.x"

      - name: Run planning gate check
        run: python scripts/ensure_planning.py
```

---

# Optional marker block for `TECH-SPEC.md`

Add this near the top of `TECH-SPEC.md` so the script has clean markers to check:

```md
## Machine-Readable Stack Markers

Runtime:
Language:
Database:
Dependencies:
```

---

# Optional marker block for `TASKS.md`

Add this near the top of `TASKS.md`:

```md
FIRST BUILD SLICE:
```

File: scripts/ensure_planning.js
#!/usr/bin/env node
/**
 * Planning gate checker for ai-project-starter.
 * * This script intentionally uses simple marker checks.
 * It does not use fuzzy AI validation.
 * * It fails when required planning files are missing, empty, or still unresolved.
 */

const fs = require('fs');
const path = require('path');

// Replicates Path(__file__).resolve().parents[1] assuming script lives in /scripts
const ROOT = path.resolve(__dirname, '..');

const REQUIRED_FILES = [
    "PROJECT-INTAKE.md",
    "AGENTS.md",
    "HARNESS-LOOP.md",
    "00-PLANNING/DECISIONS-TO-MAKE.md",
    "ARCHITECTURE.md",
    "TECH-SPEC.md",
    "TASKS.md",
    "AI-NOTES.md",
];

const OPTIONAL_CONTEXT_FILES = [
    "DATABASE.md",
    "DATA-FLOW.md",
];

const PLACEHOLDER_PATTERNS = [
    /\bTODO\b/i,
    /\bTBD\b/i,
    /\bFILL\s+THIS\b/i,
    /\bPLACEHOLDER\b/i,
    /\bUNRESOLVED\b/i,
    /_Describe/i,
    /_One paragraph/i,
    /___________/i,
];

function readFile(filePath) {
    try {
        return fs.readFileSync(filePath, 'utf8');
    } catch (error) {
        return "";
    }
}

function fail(message, failures) {
    failures.push(`- ${message}`);
}

function fileHasRealContent(content, minNonEmptyLines = 5) {
    const lines = content.split(/\r?\n/).map(line => line.trim()).filter(line => line.length > 0);
    return lines.length >= minNonEmptyLines;
}

function containsBadPlaceholder(content) {
    const found = [];
    for (const pattern of PLACEHOLDER_PATTERNS) {
        if (pattern.test(content)) {
            // Source extracts the text representation of the regex pattern
            found.push(`\\b${pattern.source}\\b`);
        }
    }
    return found;
}

function checkRequiredFiles(failures) {
    const missingFiles = [];

    for (const relativePath of REQUIRED_FILES) {
        const filePath = path.join(ROOT, relativePath);

        if (!fs.existsSync(filePath)) {
            fail(`Missing required file: \`${relativePath}\``, failures);
            missingFiles.push(relativePath);
            continue;
        }

        const content = readFile(filePath);

        if (!fileHasRealContent(content)) {
            fail(`\`${relativePath}\` appears empty or too thin.`, failures);
            continue;
        }

        const placeholders = containsBadPlaceholder(content);
        if (placeholders.length > 0) {
            fail(`\`${relativePath}\` still contains placeholder markers: ${placeholders.join(', ')}`, failures);
        }
    }

    return missingFiles.length > 0;
}

function checkProjectIntake(failures) {
    const filePath = path.join(ROOT, "PROJECT-INTAKE.md");
    if (!fs.existsSync(filePath)) return;

    const content = readFile(filePath);
    if (!content.includes("COMPLETED: true")) {
        fail(`\`PROJECT-INTAKE.md\` must include \`COMPLETED: true\` before coding.`, failures);
    }
}

function checkDecisionGates(failures) {
    const filePath = path.join(ROOT, "00-PLANNING/DECISIONS-TO-MAKE.md");
    if (!fs.existsSync(filePath)) return;

    const content = readFile(filePath);
    const unresolvedMarkers = [
        "STATUS: UNRESOLVED",
        "STATUS: NEEDS_SHAUN",
        "STATUS: NEEDS_SENIOR_DEVELOPER",
        "NEEDS SHAUN",
        "NEEDS SENIOR DEVELOPER",
    ];

    for (const marker of unresolvedMarkers) {
        if (content.includes(marker)) {
            fail(`\`DECISIONS-TO-MAKE.md\` still contains blocker marker: \`${marker}\``, failures);
        }
    }

    // Dynamic global match count to replicate python content.count()
    const resolvedMatches = content.match(/STATUS: RESOLVED/g);
    const resolvedCount = resolvedMatches ? resolvedMatches.length : 0;

    if (resolvedCount < 5) {
        fail(
            `\`DECISIONS-TO-MAKE.md\` should contain at least 5 resolved gates: ` +
            `Project type, Architecture shape, Stack choice, Database choice, Data flow (or alternatives).`,
            failures
        );
    }
}

function checkTechSpec(failures) {
    const filePath = path.join(ROOT, "TECH-SPEC.md");
    if (!fs.existsSync(filePath)) return;

    const content = readFile(filePath);
    const requiredMarkers = [
        "Runtime:",
        "Language:",
        "Database:",
        "Dependencies:",
    ];

    for (const marker of requiredMarkers) {
        if (!content.includes(marker)) {
            fail(`\`TECH-SPEC.md\` should include \`${marker}\` in Machine-Readable Stack Markers.`, failures);
        }
    }
}

function checkArchitecture(failures) {
    const filePath = path.join(ROOT, "ARCHITECTURE.md");
    if (!fs.existsSync(filePath)) return;

    const content = readFile(filePath);
    const requiredMarkers = [
        "System Overview",
        "Architecture Style",
        "Directory Structure",
    ];

    for (const marker of requiredMarkers) {
        if (!content.includes(marker)) {
            fail(`\`ARCHITECTURE.md\` should include \`${marker}\`.`, failures);
        }
    }
}

function checkTaskSlice(failures) {
    const filePath = path.join(ROOT, "TASKS.md");
    if (!fs.existsSync(filePath)) return;

    const content = readFile(filePath);
    if (!content.includes("FIRST BUILD SLICE:")) {
        fail(`\`TASKS.md\` must include \`FIRST BUILD SLICE:\` before coding.`, failures);
    }
}

function checkContextFiles(failures) {
    const archPath = path.join(ROOT, "ARCHITECTURE.md");
    if (!fs.existsSync(archPath)) return;

    const archContent = readFile(archPath).toLowerCase();

    if (archContent.includes("database") && !fs.existsSync(path.join(ROOT, "DATABASE.md"))) {
        fail(`\`ARCHITECTURE.md\` mentions database but \`DATABASE.md\` is missing.`, failures);
    }

    if (archContent.includes("data flow") && !fs.existsSync(path.join(ROOT, "DATA-FLOW.md"))) {
        fail(`\`ARCHITECTURE.md\` mentions data movement but \`DATA-FLOW.md\` is missing.`, failures);
    }
}

function printFailureInstructions(failures) {
    console.log("Planning check failed.\n");
    console.log("Fix these items before coding or merging:\n");
    console.log(failures.join("\n"));
    console.log("\nNext steps:");
    console.log("1. Read ONBOARDING.md for the two-phase workflow and planning checklist");
    console.log("2. Fill the missing or incomplete files listed above");
    console.log("3. Push your changes to trigger the check again");
    console.log("4. Run: node scripts/ensure_planning.js (locally to verify)");
}

function main() {
    const failures = [];

    // Core file assertions
    const coreFilesMissing = checkRequiredFiles(failures);

    // Early exit if the foundational layout is physically missing
    if (coreFilesMissing || failures.length > 0) {
        printFailureInstructions(failures);
        process.exit(1);
    }

    // Run structural inner gate verifications
    checkProjectIntake(failures);
    checkDecisionGates(failures);
    checkTechSpec(failures);
    checkArchitecture(failures);
    checkTaskSlice(failures);
    checkContextFiles(failures);

    if (failures.length > 0) {
        printFailureInstructions(failures);
        process.exit(1);
    }

    console.log("Planning check passed.");
    process.exit(0);
}

// Kick off run
main();
